### PR TITLE
Fix to #30244 - Json column fails to materialize only when default interceptor used

### LINKED ISSUE #30244: Json column fails to materialize only when default interceptor used
I was trying to implement a materialization interceptor when I came across this issue. The attached project implements the interceptor but just as a pass through. The interceptor doesn't have to do anything to cause the error. 

I do receive the same type of error as #29380 but I believe this is different as my Json doesn't use Id.

Attaching project that reproduces the issue. Run dotnet ef database update. Run application and watch it output the values in the json column. Uncomment line 14 in the TestDbContext and run the application again and get the following stack trace...

```
   at System.Linq.Expressions.Expression.ValidateLambdaArgs(Type delegateType, Expression& body, ReadOnlyCollection`1 parameters, String paramName)
   at System.Linq.Expressions.Expression.Lambda[TDelegate](Expression body, String name, Boolean tailCall, IEnumerable`1 parameters)
   at System.Linq.Expressions.Expression.Lambda[TDelegate](Expression body, Boolean tailCall, IEnumerable`1 parameters)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitNew(NewExpression node)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitBlockExpressions(ExpressionVisitor visitor, BlockExpression block)
   at System.Linq.Expressions.ExpressionVisitor.VisitBlock(BlockExpression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitBinary(BinaryExpression node)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.VisitBinary(BinaryExpression binaryExpression)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitBlockExpressions(ExpressionVisitor visitor, BlockExpression block)
   at System.Linq.Expressions.ExpressionVisitor.VisitBlock(BlockExpression node)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitBlockExpressions(ExpressionVisitor visitor, BlockExpression block)
   at System.Linq.Expressions.ExpressionVisitor.VisitBlock(BlockExpression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitSwitchCase(SwitchCase node)
   at System.Linq.Expressions.ExpressionVisitor.Visit[T](ReadOnlyCollection`1 nodes, Func`2 elementVisitor)
   at System.Linq.Expressions.ExpressionVisitor.VisitSwitch(SwitchExpression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitBinary(BinaryExpression node)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.VisitBinary(BinaryExpression binaryExpression)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitBlockExpressions(ExpressionVisitor visitor, BlockExpression block)
   at System.Linq.Expressions.ExpressionVisitor.VisitBlock(BlockExpression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitConditional(ConditionalExpression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitConditional(ConditionalExpression node)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitBlockExpressions(ExpressionVisitor visitor, BlockExpression block)
   at System.Linq.Expressions.ExpressionVisitor.VisitBlock(BlockExpression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitBinary(BinaryExpression node)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.VisitBinary(BinaryExpression binaryExpression)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitBlockExpressions(ExpressionVisitor visitor, BlockExpression block)
   at System.Linq.Expressions.ExpressionVisitor.VisitBlock(BlockExpression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitLambda[T](Expression`1 node)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.VisitExtension(Expression extensionExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.ProcessShaper(Expression shaperExpression, RelationalCommandCache& relationalCommandCache, IReadOnlyList`1& readerColumns, LambdaExpression& relatedDataLoaders, Int32& collectionId)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.VisitShapedQuery(ShapedQueryExpression shapedQueryExpression)
   at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.VisitExtension(Expression extensionExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.VisitExtension(Expression extensionExpression)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass9_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at Program.<Main>$(String[] args) in D:\Source\repos\FMB.Net\IssueRepro\Program.cs:line 5
```
EF Core version: 7.0.2
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 7.0
Operating system: Windows 10 Pro
IDE: Visual Studio 2022 17.4.4
[IssueRepro.zip](https://github.com/dotnet/efcore/files/10700843/IssueRepro.zip)

### PR DESCRIPTION
**Pull request by maumar**
Problem was that when we build materializer expression, ValueBufferTryReadValue call that extracts values of keys is typed with key's CLR type. This fails for json as we visit the materializer expression, because calls to ValueBufferTryReadValue are replaced with array accesses to object[] that we store key values in. We should be using ValueBufferTryReadValue<object> when trying to get value for keys, like we do everywhere else, so that after the visitation expression types are correct.

Fixes #30244

### CODE REVIEW

File: test/EFCore.Relational.Specification.Tests/Query/JsonQueryAdHocTestBase.cs [hunk 1]
```diff
@@ -231,6 +231,87 @@ public virtual async Task Missing_navigation_works_with_deduplication(bool async
 
     #endregion
 
+    #region 30244
+
+    protected virtual void OnConfiguring30244(DbContextOptionsBuilder builder)
+    {
+        builder.AddInterceptors(new IssueReproMaterialization30244());
```
Comment (maumar) [inline test/EFCore.Relational.Specification.Tests/Query/JsonQueryAdHocTestBase.cs:238]: how to properly incorporate interceptors into the test infra? @ajcvickers

Comment (maumar) [reply test/EFCore.Relational.Specification.Tests/Query/JsonQueryAdHocTestBase.cs:238]: i'm getting: A call was made to 'AddInterceptors', but Entity Framework is not building its own internal service provider. Either allow Entity Framework to build the service provider by removing the call to 'UseInternalServiceProvider', or build the 'ISingletonInterceptor' services to use into the service provider before passing it to 'UseInternalServiceProvider'.

**Comment by ajcvickers** [comment]
@maumar I moved the testing to the interceptor tests, which are setup to handle this.

File: test/EFCore.Sqlite.FunctionalTests/MaterializationInterceptionSqliteTest.cs [hunk 2]
```diff
@@ -25,5 +43,8 @@ protected override ITestStoreFactory TestStoreFactory
             IServiceCollection serviceCollection,
             IEnumerable<ISingletonInterceptor> injectedInterceptors)
             => base.InjectInterceptors(serviceCollection.AddEntityFrameworkSqlite(), injectedInterceptors);
+
+        public override DbContextOptionsBuilder AddOptions(DbContextOptionsBuilder builder)
+            => base.AddOptions(builder.ConfigureWarnings(e => e.Ignore(SqliteEventId.CompositeKeyWithValueGeneration)));
```
Comment (ajcvickers) [inline test/EFCore.Sqlite.FunctionalTests/MaterializationInterceptionSqliteTest.cs:48]: @maumar Had to suppress his warning, which I don't think should apply when the owned type is mapped to JSON. Are we tracking this somewhere?

**Review by ajcvickers — APPROVED** [review]

File: test/EFCore.Sqlite.FunctionalTests/MaterializationInterceptionSqliteTest.cs (hunk 2 above)
Comment (maumar) [reply test/EFCore.Sqlite.FunctionalTests/MaterializationInterceptionSqliteTest.cs:48]: https://github.com/dotnet/efcore/issues/26708, but maybe we need something json specific?

Comment (ajcvickers) [reply test/EFCore.Sqlite.FunctionalTests/MaterializationInterceptionSqliteTest.cs:48]: Will people get this warning if they explicitly configure a composite key for a mapped JSON collection when that configuration is valid? /cc @AndriySvyryd

Comment (AndriySvyryd) [reply test/EFCore.Sqlite.FunctionalTests/MaterializationInterceptionSqliteTest.cs:48]: > Will people get this warning if they explicitly configure a composite key for a mapped JSON collection when that configuration is valid?

Yes, unless they mark the non-fk property as never generated

Comment (maumar) [reply test/EFCore.Sqlite.FunctionalTests/MaterializationInterceptionSqliteTest.cs:48]: https://github.com/dotnet/efcore/issues/30516

**Pull request merged**